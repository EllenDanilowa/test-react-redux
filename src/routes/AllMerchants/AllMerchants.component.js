import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {Link} from 'react-router-dom';
import Merchant from '../../components/Merchant/Merchant';
import {Header, Title, PageWrapper, Icon} from './AllMerchants.styled';
import PlusIcon from './assets/plus.svg';

class AllMerchants extends Component {
  render() {
    const {loading, error, merchants, deleteMerchant} = this.props;

    return (
      <PageWrapper>
        <Header>
          <Title>All merchants</Title>
          <div>
            <Link to='/add'>
              <Icon src={PlusIcon}/>
                Add new merchant
            </Link>
          </div>
        </Header>
        <div>
          {!(error || loading) && merchants && merchants.map((merchant) =>
            <Merchant
              key={merchant.id}
              item={merchant}
              delete={deleteMerchant}
            />
          )}
          {loading && <p>Loading...</p>}
          {error && <p>Sorry, we could not load data</p>}
        </div>
      </PageWrapper>
    );
  }
}

AllMerchants.propTypes = {
  loading: PropTypes.bool,
  error: PropTypes.bool,
  merchants: PropTypes.array,
  deleteMerchant: PropTypes.func.isRequired
};

export default AllMerchants;
