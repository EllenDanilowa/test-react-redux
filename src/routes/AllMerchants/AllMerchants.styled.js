import styled from 'styled-components';
import colors from '../../styles/colors';

export const Header = styled.div`
  align-items: center;
  display: flex;
  justify-content: space-between;
  flex-wrap: wrap;
`;

// TODO: create a common place for title
export const Title = styled.h3`
  color: ${colors.title};
  font-size: 32px;
  font-weight: 700;
`;

export const PageWrapper = styled.div`
  margin: 12px;
`;

export const Icon = styled.img`
  padding-right: 6px;
  vertical-align: text-top;
  width: 16px;
`;
