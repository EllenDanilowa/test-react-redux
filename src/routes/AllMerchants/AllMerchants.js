import {connect} from 'react-redux';
import AllMerchants from './AllMerchants.component';
import {deleteMerchant} from '../../redux/merchant/merchant.actions';

const mapStateToProps = (state) => ({
  loading: state.merchant.loading,
  error: state.merchant.error,
  merchants: state.merchant.items,
});

const mapDispatchToProps = (dispatch) => ({
  deleteMerchant: (id) => {
    dispatch(deleteMerchant(id));
  }
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AllMerchants);
