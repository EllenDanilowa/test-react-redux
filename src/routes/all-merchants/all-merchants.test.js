import React from 'react';
import renderer from 'react-test-renderer';
import AllMerchants from "./all-merchants.component";
import {MemoryRouter} from 'react-router';

jest.mock('../../components/merchant-list/merchant-list', () => () => 'Merchant List component');

describe('AllMerchants', () => {
  let element;
  let merchants;
  let merchantsCount;
  let deleteMerchant;
  let fetchMerchants;
  let updateVisibleMerchants;

  const createElement = (props) => {
    const template = (<MemoryRouter><AllMerchants {...props}/></MemoryRouter>);

    return renderer.create(template).toJSON();
  };

  beforeEach(() => {
    merchantsCount = 0;
    deleteMerchant = () => {};
    updateVisibleMerchants = () => {};
    fetchMerchants = jest.fn();
    merchants = [];

    jest.spyOn(React, 'useEffect').mockImplementation((cb) => cb());
  });

  it('fetches merchants on mounting', () => {
    element = createElement({merchants, merchantsCount, deleteMerchant, fetchMerchants, updateVisibleMerchants});

    expect(fetchMerchants).toHaveBeenCalled();
  });

  it('renders component with loading message when loading', () => {
    const loading = true;
    element = createElement({loading, merchants, merchantsCount, deleteMerchant, fetchMerchants, updateVisibleMerchants});

    expect(fetchMerchants).toHaveBeenCalled();
    expect(element).toMatchSnapshot();
  });

  it('renders component with error message if error', () => {
    const error = true;
    element = createElement({error, merchants, merchantsCount, deleteMerchant, fetchMerchants, updateVisibleMerchants});

    expect(element).toMatchSnapshot();
  });

  it('renders component with merchant list if count is greater than 0', () => {
    merchantsCount = 5;
    element = createElement({merchants, merchantsCount, deleteMerchant, fetchMerchants, updateVisibleMerchants});

    expect(element).toMatchSnapshot();
  });

  it('renders no main content if count is equal to 0', () => {
    merchantsCount = 0;
    element = createElement({merchants, merchantsCount, deleteMerchant, fetchMerchants, updateVisibleMerchants});

    expect(element).toMatchSnapshot();
  });
});
