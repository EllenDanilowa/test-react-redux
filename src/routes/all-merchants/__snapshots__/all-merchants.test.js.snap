// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllMerchants renders component with error message if error 1`] = `
<div
  className="sc-AxirZ cSToxy"
>
  <div
    className="sc-AxiKw iuHfWT"
  >
    <h3
      className="sc-AxjAm hnoMwN"
    >
      All merchants
    </h3>
    <div>
      <a
        href="/create"
        onClick={[Function]}
      >
        <img
          className="sc-AxhCb dwZzjx"
          src="plus.svg"
        />
        Add new merchant
      </a>
    </div>
  </div>
  <div>
    <p>
      Sorry, we could not load data
    </p>
  </div>
</div>
`;

exports[`AllMerchants renders component with loading message when loading 1`] = `
<div
  className="sc-AxirZ cSToxy"
>
  <div
    className="sc-AxiKw iuHfWT"
  >
    <h3
      className="sc-AxjAm hnoMwN"
    >
      All merchants
    </h3>
    <div>
      <a
        href="/create"
        onClick={[Function]}
      >
        <img
          className="sc-AxhCb dwZzjx"
          src="plus.svg"
        />
        Add new merchant
      </a>
    </div>
  </div>
  <div>
    <p>
      Loading...
    </p>
  </div>
</div>
`;

exports[`AllMerchants renders component with merchant list if count is greater than 0 1`] = `
<div
  className="sc-AxirZ cSToxy"
>
  <div
    className="sc-AxiKw iuHfWT"
  >
    <h3
      className="sc-AxjAm hnoMwN"
    >
      All merchants
    </h3>
    <div>
      <a
        href="/create"
        onClick={[Function]}
      >
        <img
          className="sc-AxhCb dwZzjx"
          src="plus.svg"
        />
        Add new merchant
      </a>
    </div>
  </div>
  <div>
    Merchant List component
  </div>
</div>
`;

exports[`AllMerchants renders no main content if count is equal to 0 1`] = `
<div
  className="sc-AxirZ cSToxy"
>
  <div
    className="sc-AxiKw iuHfWT"
  >
    <h3
      className="sc-AxjAm hnoMwN"
    >
      All merchants
    </h3>
    <div>
      <a
        href="/create"
        onClick={[Function]}
      >
        <img
          className="sc-AxhCb dwZzjx"
          src="plus.svg"
        />
        Add new merchant
      </a>
    </div>
  </div>
  <div />
</div>
`;
